browser/ai_chat/tools/tab_management_tool.cc

```diff
@@ -0,0 +1,1378 @@
+  // TODO: keep sidebar open to a AI Chat conversation if it's open in the
+  // active window so that user can follow along. Or open in a full page if
+  // we're going to remove the active window with the active conversation.
+  if (client_data && client_data->is_bool()) {
+    bool granted = client_data->GetBool();
+    if (!granted) {
+      // Let the AI know, so that the conversation can proceed. The AI
+      // might have asked to use this tool and the user did not expect
+      // it.
+      std::move(callback).Run(CreateContentBlocksForText(
+          "User has not granted permission to use this tool. Perhaps the use "
+          "of this tool was not expected "
+          "by the user."));
+      return;
+    }
+    user_has_granted_permission_ = true;
+  }
+
+  CHECK(user_has_granted_permission_)
+      << "This tool should not be called unless by the client with the correct "
+         "permission format, or automatically after "
+         "RequiresUserInteractionBeforeHandling() "
+         "returns false";
+
+  data_decoder::DataDecoder::ParseJsonIsolated(
+      input_json,
+      base::BindOnce(&TabManagementTool::OnParseJson,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback)));
+}
+
+void TabManagementTool::OnParseJson(
+    UseToolCallback callback,
+    data_decoder::DataDecoder::ValueOrError result) {
+  if (!result.has_value() || !result->is_dict()) {
+    std::move(callback).Run(
+        CreateContentBlocksForText("Failed to parse input JSON. Please provide "
+                                   "valid JSON with an 'action' field."));
+    return;
+  }
+
+  const auto& dict = result->GetDict();
+  const auto* action = dict.FindString("action");
+
+  if (!action) {
+    std::move(callback).Run(CreateContentBlocksForText(
+        "Missing required 'action' field. Must be one of: list, move, close, "
+        "create_group, update_group, remove_from_group"));
+    return;
+  }
+
+  if (*action == "list") {
+    HandleListTabs(std::move(callback));
+  } else if (*action == "move") {
+    HandleMoveTabs(std::move(callback), dict);
+  } else if (*action == "close") {
+    HandleCloseTabs(std::move(callback), dict);
+  } else if (*action == "create_group") {
+    HandleCreateGroup(std::move(callback), dict);
+  } else if (*action == "update_group") {
+    HandleUpdateGroup(std::move(callback), dict);
+  } else if (*action == "remove_from_group") {
+    HandleRemoveFromGroup(std::move(callback), dict);
+  } else {
+    std::move(callback).Run(CreateContentBlocksForText(
+        "Invalid action. Must be one of: list, move, close, "
+        "create_group, update_group, remove_from_group"));
+  }
+}
+
+base::Value::Dict TabManagementTool::GenerateTabList() const {
+  base::Value::Dict result;
+  base::Value::List windows;
+
+  // Iterate through all browser windows for this profile
+  for (Browser* browser : *BrowserList::GetInstance()) {
```

Comment:
Hey, so I tried this approach initially for the `TabTrackerService`, but we're trying to avoid accessing globals (like BrowserList and Browser).

Unfortunately you might need to extend the TabTrackerService to track windows and groups instead

Reply:
Maybe let's extend it in a separate PR which lands before this one.

Reply:
And if we do that, maybe it'd make more sense to move the `Removed/Insertion` waiters there too?